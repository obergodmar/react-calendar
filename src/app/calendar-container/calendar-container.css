.calendar {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    width: 224px;
    height: 274px;
    padding: 13px;
    font-size: 12px;
    background-color: #fff;
    border: 1px solid gray;
    box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
    user-select: none;
}

.todayButton {
    align-self: flex-end;
    margin-top: 5px;
    margin-right: 7px;
    color: #0078d4;
    font-size: 13px;
    background-color: #fff;
    border: none;
}

.todayButton:hover {
    background-color: #b2d5ff;
}
