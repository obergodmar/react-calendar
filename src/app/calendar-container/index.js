export { CalendarContainer } from "./calendar-container";
