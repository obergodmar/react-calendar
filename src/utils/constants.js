export const TEST_DATE = new Date(1970, 1, 1);
export const FULL_FORMAT = "MMddyyyy";
export const WEEKDAY_FORMAT = "EEEEEE";
export const MONTH_FORMAT = "MMM";
export const MONTH_YEAR = "LLLL yyyy";
export const DATE_FORMAT = "MM / dd / yyyy";
