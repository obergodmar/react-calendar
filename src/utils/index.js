export { default as dayMatrix } from "./day-matrix";
export * from "./constants";
