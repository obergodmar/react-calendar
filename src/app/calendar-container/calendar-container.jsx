import React, { Fragment, useCallback, useMemo, useState } from "react";
import PropTypes from "prop-types";
import set from "date-fns/set";
import format from "date-fns/format";
import addMonths from "date-fns/addMonths";
import addYears from "date-fns/addYears";

import { dayMatrix, MONTH_YEAR } from "../../utils";
import { MonthSelect } from "../month-select";
import { MonthSelectButton } from "../month-select-button";
import { YearPicker } from "../year-picker";
import { DayTable } from "../day-table";

import style from "./calendar-container.css";

export const CalendarContainer = ({ date, changeDate }) => {
  const [isMonthSelectShown, setMonthSelectShown] = useState(false);
  const calendar = useMemo(() => dayMatrix(date), [date]);
  const currentDate = useMemo(() => {
    const monthYear = format(date, MONTH_YEAR);
    return monthYear.charAt(0).toUpperCase() + monthYear.slice(1);
  }, [date]);

  const handleChangeDate = useCallback(
    (newDate, needClose = false) => {
      changeDate(newDate, needClose);
    },
    [changeDate]
  );

  const handleShowMonthSelect = useCallback(() => {
    setMonthSelectShown(!isMonthSelectShown);
  }, [isMonthSelectShown]);

  const handleChangeMonth = useCallback(
    (month) => {
      const newDate = set(date, { month });
      handleChangeDate(newDate);
      handleShowMonthSelect();
    },
    [date, handleChangeDate, handleShowMonthSelect]
  );

  const handleAddMonth = useCallback(
    (amount) => {
      const newDate = addMonths(date, amount);
      handleChangeDate(newDate);
    },
    [date, handleChangeDate]
  );

  const handleAddYear = useCallback(
    (amount) => {
      const newDate = addYears(date, amount);
      handleChangeDate(newDate);
    },
    [date, handleChangeDate]
  );

  const handleChangeToToday = useCallback(() => {
    handleChangeDate(new Date(), true);
  }, [handleChangeDate]);

  return (
    <div className={style.calendar}>
      <MonthSelectButton
        isShown={isMonthSelectShown}
        addMonth={handleAddMonth}
        onClick={handleShowMonthSelect}
        currentDate={currentDate}
      />
      {isMonthSelectShown ? (
        <Fragment>
          <YearPicker date={date.getFullYear()} addYear={handleAddYear} />
          <MonthSelect changeMonth={handleChangeMonth} date={date} />
        </Fragment>
      ) : (
        <Fragment>
          <DayTable calendar={calendar} changeDate={handleChangeDate} />
          <button onClick={handleChangeToToday} className={style.todayButton}>
            Today
          </button>
        </Fragment>
      )}
    </div>
  );
};

CalendarContainer.propTypes = {
  date: PropTypes.instanceOf(Date).isRequired,
  changeDate: PropTypes.func.isRequired,
};

CalendarContainer.displayName = "CalendarContainer";
