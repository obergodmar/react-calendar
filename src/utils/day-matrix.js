import { addDays, format, startOfWeek } from "date-fns";
import { FULL_FORMAT } from "./constants";

const rows = 6;
const cols = 7;
const length = rows * cols;

export default (currentDate) => {
  const today = format(new Date(), FULL_FORMAT);
  const currentDateFormat = format(currentDate, FULL_FORMAT);
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth();
  const startDate = startOfWeek(new Date(year, month, 1));
  return Array.from({ length })
    .map((_, index) => {
      const date = addDays(startDate, index);
      const dateFormat = format(date, FULL_FORMAT);
      const currentMonth = date.getMonth();
      const isPart = currentMonth === month;
      const isToday = dateFormat === today;
      const isSelected = currentDateFormat === dateFormat;
      return {
        date,
        isToday,
        isSelected,
        isPart,
      };
    })
    .reduce(
      (matrix, _, index, days) =>
        !(index % cols !== 0)
          ? [...matrix, days.slice(index, index + cols)]
          : matrix,
      []
    );
};
