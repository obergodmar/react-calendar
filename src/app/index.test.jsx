import React from "react";
import { mount, shallow } from "enzyme";
import format from "date-fns/format";
import { describe, jest, test } from "@jest/globals";

import { CalendarInput } from ".";
import { CalendarContainer } from "./calendar-container";
import { DATE_FORMAT, TEST_DATE } from "../utils";

describe("Calendar app", () => {
  const handleChange = jest.fn();
  const handleOnOpen = jest.fn();
  const date = new Date(TEST_DATE);

  const wrapper = shallow(
    <CalendarInput value={date} onChange={handleChange} onOpen={handleOnOpen} />
  );

  test("Should match snapshot", () => {
    expect(wrapper.find(CalendarContainer)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  test("Should render calendar and match snapshot", () => {
    wrapper.find(".input").simulate("click");

    expect(wrapper.find(CalendarContainer)).toHaveLength(1);
    expect(handleOnOpen).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  test("Should hide calendar and trigger onClose", () => {
    const handleOnClose = jest.fn();
    const component = shallow(
      <CalendarInput
        isOpened
        value={date}
        onChange={handleChange}
        onClose={handleOnClose}
      />
    );

    component.find(".input").simulate("click");
    expect(handleOnClose).toHaveBeenCalled();
    expect(component.find(CalendarContainer)).toHaveLength(0);
  });

  test("Should set today's date if date prop is not valid", () => {
    const component = shallow(
      <CalendarInput value={null} onChange={handleChange} />
    );

    expect(component.find(".input").props().children).toEqual(
      format(new Date(), DATE_FORMAT)
    );
  });

  test("Should change date", () => {
    const mockHandleChange = jest.fn();
    const component = shallow(
      <CalendarInput isOpened onChange={mockHandleChange} value={date} />
    );

    component.find(CalendarContainer).props().changeDate(new Date(), false);
    expect(mockHandleChange).toHaveBeenCalled();

    jest.useFakeTimers();
    jest.runOnlyPendingTimers();
    component.update();
    expect(component.find(".calendar")).toHaveLength(1);
  });

  test("Should change date and close", () => {
    const mockHandleChange = jest.fn();
    const component = shallow(
      <CalendarInput isOpened onChange={mockHandleChange} value={date} />
    );

    component.find(CalendarContainer).props().changeDate(new Date(), true);
    expect(mockHandleChange).toHaveBeenCalled();

    jest.useFakeTimers();
    jest.runOnlyPendingTimers();
    component.update();
    expect(component.find(".calendar")).toEqual({});
  });

  test("Should set correct z-index", () => {
    const zIndex = 15;
    const component = shallow(
      <CalendarInput
        isOpened
        onChange={handleChange}
        value={date}
        zIndex={zIndex}
      />
    );

    expect(component.find(".calendar").props().style.zIndex).toEqual(zIndex);
  });

  test("Should be on bottom", () => {
    const component = shallow(
      <CalendarInput isOpened onChange={handleChange} value={date} />
    );

    expect(component.find(".calendar").props().style.top).toEqual(0);
    expect(component.find(".calendar").props().style.transform).toEqual(
      "translateY(0)"
    );
  });

  test("Should trigger scroll listener", () => {
    window.addEventListener = jest.fn();
    mount(<CalendarInput isOpened onChange={handleChange} value={date} />);

    expect(window.addEventListener).toHaveBeenCalled();
  });

  test("Should be on top", () => {
    Element.prototype.getBoundingClientRect = jest.fn(() => ({
      bottom: 0,
      height: 50,
      left: 0,
      right: 0,
      top: 500,
      width: 0,
      x: 0,
      y: 0,
      toJSON: () => {},
    }));
    const component = mount(
      <CalendarInput isOpened onChange={handleChange} value={date} />
    );
    window.dispatchEvent(new Event("scroll"));

    expect(component.find(".calendar").at(0).props().style.top).not.toEqual(0);
    expect(
      component.find(".calendar").at(0).props().style.transform
    ).not.toEqual("translateY(0)");
  });
});
