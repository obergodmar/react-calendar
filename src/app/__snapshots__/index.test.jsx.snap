// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar app Should match snapshot 1`] = `
<Fragment>
  <div
    className="calendarContainer"
  >
    <button
      className="input"
      onClick={[Function]}
    >
      02 / 01 / 1970
    </button>
  </div>
</Fragment>
`;

exports[`Calendar app Should render calendar and match snapshot 1`] = `
<Fragment>
  <div
    className="calendarContainer"
  >
    <button
      className="input"
      onClick={[Function]}
    >
      02 / 01 / 1970
    </button>
    <div
      className="calendar"
      style={
        Object {
          "top": 0,
          "transform": "translateY(0)",
          "transition": "transform .2s ease-out",
          "zIndex": 1,
        }
      }
    >
      <CalendarContainer
        changeDate={[Function]}
        date={1970-01-31T21:00:00.000Z}
        isOpened={false}
        onClose={[Function]}
        onOpen={[Function]}
        zIndex={1}
      />
    </div>
  </div>
</Fragment>
`;
